<a href="https://colab.research.google.com/github/scottwmwork/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [47]:
### YOUR CODE STARTS HERE
#LOAD AND CLEAN DATA
import pandas as pd
!wget https://raw.githubusercontent.com/scottwmwork/datasets/master/house-votes-84.data #uploaded from my own github repo of datasets

--2019-08-19 21:43:07--  https://raw.githubusercontent.com/scottwmwork/datasets/master/house-votes-84.data
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [text/plain]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.007s  

2019-08-19 21:43:07 (2.56 MB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [0]:
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

In [49]:
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
import numpy as np
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

In [51]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [52]:
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

In [0]:
rep = df[df.party == 'republican']
dem = df[df.party == 'democrat']

In [55]:
#2 Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
print("republicans:",rep['handicapped-infants'].mean())
print("democrat:",dem['handicapped-infants'].mean())
ttest_ind(dem['handicapped-infants'],rep['handicapped-infants'],nan_policy = "omit")

republicans: 0.18787878787878787
democrat: 0.6046511627906976


Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

### 2.) Find an issue that democrats support more than republicans with p < 0.01
Null Hypothesis: Republicans support bill just as much as Democrats


Alternative Hypothesis: Republicans do not support bill just as much as Democrats


Confidence Level: 95%


T-statistic: 9.205264294809222


P-Value: 1.613440327937243e-18

Conclusion: We reject the null hypothesis, since the P-value of 1.613440327937243e-18 is less than 0.01 


In [56]:
#3 Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
print('democrats:',dem['religious-groups'].mean())
print('republicans:',rep['religious-groups'].mean())
ttest_ind(dem['religious-groups'],rep['religious-groups'], nan_policy = "omit")

democrats: 0.47674418604651164
republicans: 0.8975903614457831


Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

###3.) Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01###

Null Hypothesis: Democrats support bill just as much as republicans

Alternative Hypothesis: Democrats do not support bill just as much as republicans

Confidence Level: 95

T-statistic:-9.737575825219457

P-Value: 2.3936722520597287e-20

Conclusion: We reject the null hypothesis, since the P-value of 2.3936722520597287e-20 is less than 0.01

In [57]:
#4 Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)
print('democrats:',dem['water-project'].mean())
print('republicans:',rep['water-project'].mean())
ttest_ind(rep['water-project'],dem['water-project'], nan_policy = "omit")

democrats: 0.502092050209205
republicans: 0.5067567567567568


Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

###4.) Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)###

Null Hypothesis: Democrats and Republicans support the bill equally

Alternative Hypothesis: Democrats and Republicans support the bill unequally

Confidence Level: 95%

T-statistic: 0.08896538137868286

P-Value: 0.9291556823993485

Conclusion: Fail to reject null hypothesis due to a p-value of 0.9291556823993485 which is greater than .01